packages/json-rpc-provider-proxy/src/get-sync-provider.ts

```diff
@@ -0,0 +1,132 @@
+import type { RequestId } from "./internal-types"
+import type { ConnectProvider, Provider } from "@polkadot-api/json-rpc-provider"
+import { getSubscriptionManager } from "./subscription-manager"
+
+export type ConnectAsyncProvider = (
+  onMessage: (message: string) => void,
+  onHalt: () => void,
+) => Provider
+
+export const getSyncProvider =
+  (input: () => Promise<ConnectAsyncProvider>): ConnectProvider =>
+  (onMessage) => {
+    // if it's null it means that the consumer has called `disconnect`
+    // of it's a Promise it means that it's being respolved, otherwise it's resolved
+    let provider: Provider | Promise<Provider> | null
+
+    let bufferedMessages: Array<string> = []
+    const pendingResponses = new Set<RequestId>()
+    const subscriptionManager = getSubscriptionManager()
+
+    const onMessageProxy = (message: string) => {
+      let parsed: any
+      try {
+        parsed = JSON.parse(message)
+      } catch (_) {
+        console.error(`Unable to parse incoming message: ${message}`)
+        return
+      }
+
+      if (parsed.id !== undefined) {
+        pendingResponses.delete(parsed.id)
+        subscriptionManager.onResponse(parsed)
+      } else {
+        subscriptionManager.onNotifiaction(parsed)
+      }
+
+      onMessage(message)
+    }
+
+    const send = (message: string) => {
+      if (!provider) return
+
+      const parsed = JSON.parse(message)
+      subscriptionManager.onSent(parsed)
+      if (parsed.id) pendingResponses.add(parsed.id)
+
+      if (provider instanceof Promise) {
+        bufferedMessages.push(message)
+      } else provider.send(message)
+    }
+
+    const onHalt = (): Promise<Provider> => {
+      bufferedMessages = []
+      const pendingResponsesCopy = [...pendingResponses]
+      pendingResponses.clear()
+
+      // It needs to restart before sending the errored
+      // responses/notifications because the consumer may
+      // react to those by sending new requests
+      const result = start()
+
+      subscriptionManager.onAbort().forEach((msg) => {
+        onMessage(JSON.stringify(msg))
+      })
+
+      pendingResponsesCopy.forEach((id) => {
+        onMessage(
+          JSON.stringify({
+            jsonrpc: "2.0",
+            error: { code: -32603, message: "Internal error" },
+            id,
+          }),
+        )
+      })
+
+      return result
+    }
+
+    const start = (): Promise<Provider> => {
+      const onResolve = (getProvider: ConnectAsyncProvider) => {
+        let halted = false
+        const _onHalt = () => {
```

Comment:
some [linters](https://eslint.org/docs/latest/rules/no-unused-vars#argsignorepattern) ignore unused variables prefixed with `_` underscore.

To avoid confusion, perhaps it may be better to rename to `innerOnHalt` or `onHalt_`

Reply:
AFAIK that's only for unused function arguments, which is not the case here.